In [1]:
import pandas as pd

orders = pd.read_csv(r"C:\Users\Dnyaneshwar\Downloads\orders.csv")
print(orders.head())


   order_id  user_id  restaurant_id  order_date  total_amount  \
0         1     2508            450  18-02-2023        842.97   
1         2     2693            309  18-01-2023        546.68   
2         3     2084            107  15-07-2023        163.93   
3         4      319            224  04-10-2023       1155.97   
4         5     1064            293  25-12-2023       1321.91   

                  restaurant_name  
0               New Foods Chinese  
1  Ruchi Curry House Multicuisine  
2           Spice Kitchen Punjabi  
3          Darbar Kitchen Non-Veg  
4       Royal Eatery South Indian  


In [2]:
users = pd.read_json(r"C:\Users\Dnyaneshwar\Downloads\users.json")
print(users.head())


   user_id    name       city membership
0        1  User_1    Chennai    Regular
1        2  User_2       Pune       Gold
2        3  User_3  Bangalore       Gold
3        4  User_4  Bangalore    Regular
4        5  User_5       Pune       Gold


In [3]:
import sqlite3

conn = sqlite3.connect(":memory:")
with open(r"C:\Users\Dnyaneshwar\Downloads\restaurants.sql", "r") as f:
    sql_script = f.read()
conn.executescript(sql_script)

restaurants = pd.read_sql_query("SELECT * FROM restaurants;", conn)
print(restaurants.head())


   restaurant_id restaurant_name  cuisine  rating
0              1    Restaurant_1  Chinese     4.8
1              2    Restaurant_2   Indian     4.1
2              3    Restaurant_3  Mexican     4.3
3              4    Restaurant_4  Chinese     4.1
4              5    Restaurant_5  Chinese     4.8


In [4]:
merged = orders.merge(users, on="user_id", how="left")
merged = merged.merge(restaurants, on="restaurant_id", how="left")
print(merged.head())


   order_id  user_id  restaurant_id  order_date  total_amount  \
0         1     2508            450  18-02-2023        842.97   
1         2     2693            309  18-01-2023        546.68   
2         3     2084            107  15-07-2023        163.93   
3         4      319            224  04-10-2023       1155.97   
4         5     1064            293  25-12-2023       1321.91   

                restaurant_name_x       name       city membership  \
0               New Foods Chinese  User_2508  Hyderabad    Regular   
1  Ruchi Curry House Multicuisine  User_2693       Pune    Regular   
2           Spice Kitchen Punjabi  User_2084    Chennai       Gold   
3          Darbar Kitchen Non-Veg   User_319  Bangalore       Gold   
4       Royal Eatery South Indian  User_1064       Pune    Regular   

  restaurant_name_y  cuisine  rating  
0    Restaurant_450  Mexican     3.2  
1    Restaurant_309   Indian     4.5  
2    Restaurant_107  Mexican     4.0  
3    Restaurant_224  Chinese    

In [5]:
merged.to_csv(r"C:\Users\Dnyaneshwar\Downloads\final_food_delivery_dataset.csv", index=False)


In [6]:
users = pd.read_json(r"C:\Users\Dnyaneshwar\Downloads\users.json")
print(users.head())


   user_id    name       city membership
0        1  User_1    Chennai    Regular
1        2  User_2       Pune       Gold
2        3  User_3  Bangalore       Gold
3        4  User_4  Bangalore    Regular
4        5  User_5       Pune       Gold


In [7]:
import sqlite3

# Create an in-memory database
conn = sqlite3.connect(":memory:")

# Read and execute the SQL script
with open(r"C:\Users\Dnyaneshwar\Downloads\restaurants.sql", "r") as f:
    sql_script = f.read()
conn.executescript(sql_script)

# Load restaurants table into pandas
restaurants = pd.read_sql_query("SELECT * FROM restaurants;", conn)
print(restaurants.head())


   restaurant_id restaurant_name  cuisine  rating
0              1    Restaurant_1  Chinese     4.8
1              2    Restaurant_2   Indian     4.1
2              3    Restaurant_3  Mexican     4.3
3              4    Restaurant_4  Chinese     4.1
4              5    Restaurant_5  Chinese     4.8


In [8]:
merged = orders.merge(users, on="user_id", how="left")
merged = merged.merge(restaurants, on="restaurant_id", how="left")
print(merged.head())


   order_id  user_id  restaurant_id  order_date  total_amount  \
0         1     2508            450  18-02-2023        842.97   
1         2     2693            309  18-01-2023        546.68   
2         3     2084            107  15-07-2023        163.93   
3         4      319            224  04-10-2023       1155.97   
4         5     1064            293  25-12-2023       1321.91   

                restaurant_name_x       name       city membership  \
0               New Foods Chinese  User_2508  Hyderabad    Regular   
1  Ruchi Curry House Multicuisine  User_2693       Pune    Regular   
2           Spice Kitchen Punjabi  User_2084    Chennai       Gold   
3          Darbar Kitchen Non-Veg   User_319  Bangalore       Gold   
4       Royal Eatery South Indian  User_1064       Pune    Regular   

  restaurant_name_y  cuisine  rating  
0    Restaurant_450  Mexican     3.2  
1    Restaurant_309   Indian     4.5  
2    Restaurant_107  Mexican     4.0  
3    Restaurant_224  Chinese    

In [9]:
merged.to_csv(r"C:\Users\Dnyaneshwar\Downloads\final_food_delivery_dataset.csv", index=False)


In [10]:
users = pd.read_json(r"C:\Users\Dnyaneshwar\Downloads\users.json")
print(users.head())


   user_id    name       city membership
0        1  User_1    Chennai    Regular
1        2  User_2       Pune       Gold
2        3  User_3  Bangalore       Gold
3        4  User_4  Bangalore    Regular
4        5  User_5       Pune       Gold


In [11]:
import sqlite3

# Create an in-memory database
conn = sqlite3.connect(":memory:")

# Read and execute the SQL script
with open(r"C:\Users\Dnyaneshwar\Downloads\restaurants.sql", "r") as f:
    sql_script = f.read()
conn.executescript(sql_script)

# Load restaurants table into pandas
restaurants = pd.read_sql_query("SELECT * FROM restaurants;", conn)
print(restaurants.head())


   restaurant_id restaurant_name  cuisine  rating
0              1    Restaurant_1  Chinese     4.8
1              2    Restaurant_2   Indian     4.1
2              3    Restaurant_3  Mexican     4.3
3              4    Restaurant_4  Chinese     4.1
4              5    Restaurant_5  Chinese     4.8


In [12]:
merged = orders.merge(users, on="user_id", how="left")
merged = merged.merge(restaurants, on="restaurant_id", how="left")
print(merged.head())


   order_id  user_id  restaurant_id  order_date  total_amount  \
0         1     2508            450  18-02-2023        842.97   
1         2     2693            309  18-01-2023        546.68   
2         3     2084            107  15-07-2023        163.93   
3         4      319            224  04-10-2023       1155.97   
4         5     1064            293  25-12-2023       1321.91   

                restaurant_name_x       name       city membership  \
0               New Foods Chinese  User_2508  Hyderabad    Regular   
1  Ruchi Curry House Multicuisine  User_2693       Pune    Regular   
2           Spice Kitchen Punjabi  User_2084    Chennai       Gold   
3          Darbar Kitchen Non-Veg   User_319  Bangalore       Gold   
4       Royal Eatery South Indian  User_1064       Pune    Regular   

  restaurant_name_y  cuisine  rating  
0    Restaurant_450  Mexican     3.2  
1    Restaurant_309   Indian     4.5  
2    Restaurant_107  Mexican     4.0  
3    Restaurant_224  Chinese    

In [13]:
merged.to_csv(r"C:\Users\Dnyaneshwar\Downloads\final_food_delivery_dataset.csv", index=False)


In [14]:
print(merged.groupby("membership")["order_id"].count())


membership
Gold       4987
Regular    5013
Name: order_id, dtype: int64


In [15]:
print(merged.groupby("city")["total_amount"].sum())


city
Bangalore    2206946.58
Chennai      1990513.03
Hyderabad    1889366.58
Pune         1924797.93
Name: total_amount, dtype: float64


In [16]:
print(merged["user_id"].nunique())


2883


In [17]:
print(merged.groupby("membership")["total_amount"].mean().round(2))


membership
Gold       797.15
Regular    805.16
Name: total_amount, dtype: float64


In [18]:
print(merged[merged["rating"] >= 4.5]["order_id"].count())


3374


In [19]:
gold_city = merged[merged["membership"] == "Gold"].groupby("city")["total_amount"].sum()
top_city = gold_city.idxmax()
orders_in_top_city = merged[(merged["membership"] == "Gold") & (merged["city"] == top_city)]["order_id"].count()
print(top_city, orders_in_top_city)


Chennai 1337


In [20]:
print(merged.groupby(["membership","cuisine"])["total_amount"].sum())


membership  cuisine
Gold        Chinese     977713.74
            Indian      979312.31
            Italian    1005779.05
            Mexican    1012559.79
Regular     Chinese     952790.91
            Indian      992100.27
            Italian    1018424.75
            Mexican    1072943.30
Name: total_amount, dtype: float64


In [21]:
merged["order_date"] = pd.to_datetime(merged["order_date"], dayfirst=True)
print(merged.groupby(pd.PeriodIndex(merged["order_date"], freq="Q"))["total_amount"].sum())


order_date
2023Q1    1993425.14
2023Q2    1945348.72
2023Q3    2037385.10
2023Q4    2018263.66
2024Q1      17201.50
Freq: Q-DEC, Name: total_amount, dtype: float64


In [22]:
gold_orders = merged[merged["membership"] == "Gold"]["order_id"].count()
print("Total Gold orders:", gold_orders)


Total Gold orders: 4987


In [23]:
hyd_revenue = merged[merged["city"] == "Hyderabad"]["total_amount"].sum().round()
print("Hyderabad revenue:", hyd_revenue)


Hyderabad revenue: 1889367.0


In [24]:
distinct_users = merged["user_id"].nunique()
print("Distinct users:", distinct_users)


Distinct users: 2883


In [25]:
avg_gold = merged[merged["membership"] == "Gold"]["total_amount"].mean().round(2)
print("Average Gold order value:", avg_gold)


Average Gold order value: 797.15


In [26]:
high_rating_orders = merged[merged["rating"] >= 4.5]["order_id"].count()
print("Orders with rating ≥ 4.5:", high_rating_orders)


Orders with rating ≥ 4.5: 3374


In [27]:
gold_city_revenue = merged[merged["membership"] == "Gold"].groupby("city")["total_amount"].sum()
top_gold_city = gold_city_revenue.idxmax()
orders_top_city = merged[(merged["membership"] == "Gold") & (merged["city"] == top_gold_city)]["order_id"].count()
print("Top Gold city:", top_gold_city, "| Orders:", orders_top_city)


Top Gold city: Chennai | Orders: 1337


In [28]:
combo_revenue = merged.groupby(["membership","cuisine"])["total_amount"].sum()
print(combo_revenue)


membership  cuisine
Gold        Chinese     977713.74
            Indian      979312.31
            Italian    1005779.05
            Mexican    1012559.79
Regular     Chinese     952790.91
            Indian      992100.27
            Italian    1018424.75
            Mexican    1072943.30
Name: total_amount, dtype: float64


In [29]:
merged["order_date"] = pd.to_datetime(merged["order_date"], dayfirst=True)
quarterly_revenue = merged.groupby(pd.PeriodIndex(merged["order_date"], freq="Q"))["total_amount"].sum()
print(quarterly_revenue)


order_date
2023Q1    1993425.14
2023Q2    1945348.72
2023Q3    2037385.10
2023Q4    2018263.66
2024Q1      17201.50
Freq: Q-DEC, Name: total_amount, dtype: float64


# Food Delivery Hackathon Analysis
This notebook contains my submission for the Food Delivery Hackathon.  
I merged three datasets (`orders.csv`, `users.json`, `restaurants.sql`) and performed analysis to answer the hackathon questions.


## Step 1: Load Datasets
We load the three datasets into Pandas:
- Orders (CSV)
- Users (JSON)
- Restaurants (SQL)


In [35]:
import pandas as pd
orders = pd.read_csv("orders.csv")
print(orders.head())


   order_id  user_id  restaurant_id  order_date  total_amount  \
0         1     2508            450  18-02-2023        842.97   
1         2     2693            309  18-01-2023        546.68   
2         3     2084            107  15-07-2023        163.93   
3         4      319            224  04-10-2023       1155.97   
4         5     1064            293  25-12-2023       1321.91   

                  restaurant_name  
0               New Foods Chinese  
1  Ruchi Curry House Multicuisine  
2           Spice Kitchen Punjabi  
3          Darbar Kitchen Non-Veg  
4       Royal Eatery South Indian  


In [37]:
users = pd.read_json("users.json")
print(users.head())


   user_id    name       city membership
0        1  User_1    Chennai    Regular
1        2  User_2       Pune       Gold
2        3  User_3  Bangalore       Gold
3        4  User_4  Bangalore    Regular
4        5  User_5       Pune       Gold


In [38]:
import sqlite3

conn = sqlite3.connect(":memory:")
with open("restaurants.sql", "r") as f:
    sql_script = f.read()
conn.executescript(sql_script)

restaurants = pd.read_sql_query("SELECT * FROM restaurants;", conn)
print(restaurants.head())


   restaurant_id restaurant_name  cuisine  rating
0              1    Restaurant_1  Chinese     4.8
1              2    Restaurant_2   Indian     4.1
2              3    Restaurant_3  Mexican     4.3
3              4    Restaurant_4  Chinese     4.1
4              5    Restaurant_5  Chinese     4.8


## Step 2: Merge Datasets
We merge the datasets on `user_id` and `restaurant_id` to create a unified dataset.


In [39]:
merged = orders.merge(users, on="user_id", how="left")
merged = merged.merge(restaurants, on="restaurant_id", how="left")
print(merged.head())


   order_id  user_id  restaurant_id  order_date  total_amount  \
0         1     2508            450  18-02-2023        842.97   
1         2     2693            309  18-01-2023        546.68   
2         3     2084            107  15-07-2023        163.93   
3         4      319            224  04-10-2023       1155.97   
4         5     1064            293  25-12-2023       1321.91   

                restaurant_name_x       name       city membership  \
0               New Foods Chinese  User_2508  Hyderabad    Regular   
1  Ruchi Curry House Multicuisine  User_2693       Pune    Regular   
2           Spice Kitchen Punjabi  User_2084    Chennai       Gold   
3          Darbar Kitchen Non-Veg   User_319  Bangalore       Gold   
4       Royal Eatery South Indian  User_1064       Pune    Regular   

  restaurant_name_y  cuisine  rating  
0    Restaurant_450  Mexican     3.2  
1    Restaurant_309   Indian     4.5  
2    Restaurant_107  Mexican     4.0  
3    Restaurant_224  Chinese    

## Step 3: Analysis
We answer the hackathon questions using Pandas queries.


## Step 4: Hackathon MCQ Answers

1. City with highest revenue from Gold members → **Chennai**  
2. Cuisine with highest average order value → **Mexican**  
3. Distinct users with > ₹1000 total orders → **>2000**  
4. Rating range with highest revenue → **4.6 – 5.0**  
5. Gold members’ highest average order value city → **Chennai**  
6. Cuisine with lowest restaurants but strong revenue → **Chinese**  
7. % of orders by Gold members → **50%**  
8. Restaurant with highest avg order value (<20 orders) → **Grand Restaurant South Indian**  
9. Highest revenue combo → **Gold + Italian cuisine**  
10. Quarter with highest revenue → **Q3 (Jul–Sep)**


## Step 5: Hackathon Numerical Answers
**Q1: Total orders placed by Gold members**  
Answer: 4987

**Q2: Total revenue from Hyderabad city**  
Answer: 1,889,367

**Q3: Distinct users who placed at least one order**  
Answer: 2883

**Q4: Average order value for Gold members**  
Answer: 797.15

**Q5: Orders placed for restaurants with rating ≥ 4.5**  
Answer: 3374

**Q6: Orders in top revenue city among Gold members only**  
Answer: Chennai → 1337 orders

**Q7: Membership + Cuisine revenue (highest combo)**  
Answer: Gold + Italian cuisine

**Q8: Quarterly revenue (highest quarter)**  
Answer: Q3 (Jul–Sep)


## Step 6: Hackathon Fill in the Blanks Answers
**Q1: The column used to join orders.csv and users.json is**  
Answer: user_id

**Q2: The dataset containing cuisine and rating information is stored in**  
Answer: SQL

**Q3: The total number of rows in the final merged dataset is**  
Answer: 10,000

**Q4: If a user has no matching record in users.json, the merged values will be**  
Answer: null

**Q5: The Pandas function used to combine datasets based on a key is**  
Answer: merge

**Q6: The column membership in the final dataset originates from the**  
Answer: users.json file

**Q7: The join key used to combine orders data with restaurant details is**  
Answer: restaurant_id

**Q8: The column that helps identify the type of food served by a restaurant is**  
Answer: cuisine

**Q9: If a user places multiple orders, their personal details appear**  
Answer: multiple times


## Step 7: Conclusion
This hackathon helped me practice:
- Loading multiple file formats (CSV, JSON, SQL)
- Merging datasets in Pandas
- Performing groupby and aggregation queries
- Saving final results to CSV
